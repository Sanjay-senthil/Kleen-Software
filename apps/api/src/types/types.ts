import {
  AccessControlPermission,
  Axis,
  EntityList,
  FilterQuery,
  GetListingDataResults,
  GetWidgetDataResult,
  ListItemBase as ListItem,
  ListingFormatItem,
  ListingItem,
  MultiTransFormationResults,
  Transformation,
  WorkflowFilter,
} from '@kleeen/types';

// Duplicated type, must unify in types library
export interface Attribute {
  aggregation?: Transformation;
  name: string;
  // TODO @carreta remove this when XORs can be aggregated [KSE3-1735]
  isXor?: boolean;
  hasMany?: boolean;
  formatType?: string;
}

export interface WorkflowFiltersResults {
  filters: WorkflowFilter[];
}

interface ActionItem {
  link?: string;
  title?: string;
  type?: string;
}

// TODO: @cafe move this and other shared types to @kleeen/types
export interface DispatchCustomActionResults {
  data: {
    actions?: ActionItem[];
    context?: any;
    customMessage?: string;
    customTitle?: string;
    functionName: string;
    success: boolean;
  };
}

export interface MultiTransFormationArgs {
  attributes?: string[];
  entity: string;
  filters: FilterQuery;
  transformations: Transformation[];
}

export interface FormatCheckResults {
  isValid: boolean;
  errors?: {
    message: string;
  }[];
}

export interface AccessControlCheckResults {
  accessLevel: AccessControlPermission;
}

export interface AuthContext {
  dataSources: Record<string, any>;
  headers: Record<string, any>;
  user: {
    email: string;
    'cognito:id': string;
  };
  token: string;
}

export {
  Axis,
  EntityList,
  GetListingDataResults,
  GetWidgetDataResult,
  ListingFormatItem,
  ListingItem,
  ListItem,
  MultiTransFormationResults,
  Transformation,
};
