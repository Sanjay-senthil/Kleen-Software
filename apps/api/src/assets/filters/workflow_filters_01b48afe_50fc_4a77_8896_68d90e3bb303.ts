export const filterSectionFilters = [
  { name: 'Entertainment', statisticalType: 'Data - Categorical - free form' },
  { name: 'Theater', statisticalType: 'Data - Categorical - free form' },
  { name: 'Pub', statisticalType: 'Data - Categorical - free form' },
  { name: 'Games', statisticalType: 'Data - Categorical - free form' },
  { name: 'BoatHouse', statisticalType: 'Data - Categorical - free form' },
  { name: 'Deprt', statisticalType: 'Data - Categorical - free form' },
  { name: 'Services', statisticalType: 'Data - Categorical - free form' },
  { name: 'Tourism', statisticalType: 'Data - Categorical - free form' },
  { name: 'Entertainment', statisticalType: 'Data - Categorical - free form' },
  { name: 'Management', statisticalType: 'Data - Categorical' },
];
