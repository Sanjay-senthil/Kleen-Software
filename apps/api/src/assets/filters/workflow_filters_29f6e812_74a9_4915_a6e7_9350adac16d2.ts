export const filterSectionFilters = [
  { name: 'Vehicle', statisticalType: 'Data - Categorical' },
  { name: 'VehicleModel', statisticalType: 'Data - Categorical' },
  { name: 'VehicleWarranty', statisticalType: 'Data - Categorical - free form' },
  { name: 'ServiceRecord', statisticalType: 'Data - Categorical - free form' },
  { name: 'ScheduledService', statisticalType: 'Data - Categorical - free form' },
  { name: 'CurrentMileage', statisticalType: 'Data - Numeric - NTG - Discrete' },
  { name: 'CompletedService', statisticalType: 'Data - Categorical - free form' },
  { name: 'WarrantyStatus', statisticalType: 'Data - Categorical - ordered - Severity Ranking' },
  { name: 'VehicleBrand', statisticalType: 'Data - Categorical' },
];
