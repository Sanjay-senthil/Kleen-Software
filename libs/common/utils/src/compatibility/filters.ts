import {
  FilterOperation,
  FilterOperators,
  FilterQuery,
  FilterRule,
  Maybe,
  PrimitiveType,
  SingleFilterRuleValue,
  TIMESTAMP_INTERNAL_OBJECT_NAME,
} from '@kleeen/types';
import { isPrimitive, upperCamelCase } from '../helpers';

import { isNilOrEmpty } from '../validators';

interface Attribute {
  dataType?: string;
  name: string;
}

/** @deprecated Use this method to keep backward compatibility with filters generated before 10-2021, migrate to new filters version
 *
 * This method transform an object with filters generated after 10-2021 to filters with the shape of the previous version.
 *
 * @func transformFiltersToDeprecatedVersion
 * @param {DataListingArgs} input An object with a filters attribute to be transformed
 * @return {Filters} An object with filters using the previous shape
 */
export function transformFiltersToDeprecatedVersion(input: {
  attributes?: Attribute[];
  filters?: FilterQuery;
}): Maybe<Record<string, unknown>> {
  if (!input?.filters?.rules) return;

  const { attributes, filters } = input;
  const groupedRules = groupRulesByFieldOperation(filters.rules);
  const filtersInDeprecatedShape = groupedRules.reduce((acc, groupedRule) => {
    const key = getDataType(groupedRule, attributes);
    const currentGroupedRule = acc[key];

    if (!isNilOrEmpty(key)) {
      const fieldEntry = getFieldEntry(groupedRule);

      if (isPrimitive(fieldEntry)) {
        acc[key] = fieldEntry;

        return acc;
      }

      acc[key] = {
        ...currentGroupedRule,
        ...(fieldEntry as Record<string, unknown>),
      };
    }

    return acc;
  }, {});

  return filtersInDeprecatedShape;
}

//#region Private Members
/**
 * Tries to get the data type from the array of attributes,
 * if not found returns the field name capitalized as a fallback
 *
 * @param {string} fieldName to be search in the attributes list
 * @param attributes possible list of attributes to find the data type
 * @returns {string} data type id used as a filter key
 */
function getDataType(rule: FilterRule, attributes?: Attribute[]): string {
  const fieldName = rule.field;
  const attribute = attributes?.find((a) => a.name === fieldName);

  if (!attribute?.dataType) {
    if (manageAsSingleAttribute(rule)) {
      return fieldName;
    }

    return upperCamelCase(fieldName);
  }

  return attribute.dataType;
}

/**
 * Transform a rule to the previous version shape,
 * If the filter exists, tries to merge it with the corresponding transformation; for example,
 * now a timestamp rule date <= X and date >= Y has to become to:
 * date: { from: X, to: Y }
 *
 *
 * @param {FilterRule} rule to be transformed
 * @param {Record<string, unknown>} filter current filter if was previously transformed to compose accumulative changes
 * @returns {Record<string, unknown>} new filter transformed or composed from previous state
 */
function getFieldEntry(rule: FilterRule): Record<string, unknown> | PrimitiveType | undefined {
  const values = Array.isArray(rule.value) ? rule.value : [rule.value];
  const [singleValue] = values;

  switch (rule.operation) {
    case FilterOperation.GreaterOrEqualThan:
      if (rule.field === TIMESTAMP_INTERNAL_OBJECT_NAME) {
        return {
          [FilterOperators.from]: getValue(singleValue),
        };
      }

      return {
        [FilterOperators.min]: getValue(singleValue),
      };

    case FilterOperation.Is:
      if (manageAsSingleAttribute(rule)) {
        return singleValue.id;
      }

      return {
        [FilterOperators.in]: values.map(getValue),
      };

    case FilterOperation.IsNot:
      if (manageAsSingleAttribute(rule)) {
        return singleValue.id;
      }

      return {
        [FilterOperators.notIn]: values.map(getValue),
      };

    case FilterOperation.LessOrEqualThan:
      if (rule.field === TIMESTAMP_INTERNAL_OBJECT_NAME) {
        return {
          [FilterOperators.to]: getValue(singleValue),
        };
      }

      return {
        [FilterOperators.max]: getValue(singleValue),
      };

    case FilterOperation.RelativeDate:
      return {
        [FilterOperators.relativeDate]: getValue(singleValue),
      };

    default:
      throw new Error(`the rule operation ${rule.operation} is not supported.`);
  }
}

function getValue(entity?: SingleFilterRuleValue) {
  if (!entity) return;

  if (Object.prototype.hasOwnProperty.call(entity, 'displayValue')) {
    return entity.displayValue;
  }

  if (Object.prototype.hasOwnProperty.call(entity, 'id')) {
    return entity.id;
  }
}

/**
 * Verifies how to transform a rule,
 * one option is as a single value or as a complex object
 *
 * Example for a single value
 * Input :
 * {
 *   "field": "door",
 *   "operation": "=",
 *   "value": {
 *      "id": "25002565-c51b-4c90-b5e5-77d0941f483a"
 *   }
 *  }
 *
 * Output:
 * {
 *  "door": "4abd7d1b-5187-4a1d-bf63-e7a5daaaf226"
 * }
 *
 *
 * Example for a complex value
 * Input :
 * {
 *     "field": "alarmType",
 *     "operation": "=",
 *     "value": [
 *       {
 *         "displayValue": "DFO"
 *       },
 *       {
 *         "displayValue": "DHO"
 *       }
 *     ]
 *   }
 *
 * "AlarmType": {
 *   "_in": [
 *     "DFO",
 *     "DHO"
 *   ]
 * }
 *
 * @param {FilterRule} rule to be transformed
 * @returns true if the transformation has to be as a single value, otherwise as a complex object
 */
function manageAsSingleAttribute(rule: FilterRule): boolean {
  const isSingleValue = !Array.isArray(rule.value);

  if (isSingleValue && Object.prototype.hasOwnProperty.call(rule.value, 'id')) {
    return true;
  }

  return false;
}

/**
 * Simplifies a set of rules by grouping them by field and making the value an array
 * when there are multiple rules of the same field
 *
 * Example
 *
 * [
  {
    field: 'alarmType',
    value: {
      displayValue: 'DFO',
    },
  }
  {
    field: 'alarmType',
    value: {
      displayValue: 'DHO',
    },
  }
  {
    field: 'door',
    value: {
      id: '4abd7d1b-5187-4a1d-bf63-e7a5daaaf226',
    },
  }
]

=>

[
  {
    field: 'alarmType',
    value: [
      {
        displayValue: 'DFO',
      },
      {
        displayValue: 'DHO',
      },
    ]
  },
  {
    field: 'player',
    value: {
      id: '4abd7d1b-5187-4a1d-bf63-e7a5daaaf226',
    },
  }
]
 * @param rules Source rules that can contain repeated fields
 * @returns The same rules but grouped by field
 */
function groupRulesByFieldOperation(rules: FilterRule[]): FilterRule[] {
  const groupedRulesMap = rules.reduce((acc, rule) => {
    const currentRuleKey = `${rule.field}-${rule.operation}`;
    const currentRule = acc[currentRuleKey];

    if (currentRule) {
      if (!Array.isArray(currentRule.value)) {
        const value = [
          {
            ...currentRule.value,
          },
          {
            ...rule.value,
          },
        ];

        acc[currentRuleKey] = {
          ...rule,
          value,
        };

        return acc;
      } else {
        acc[currentRuleKey] = {
          ...rule,
          value: [...currentRule, rule.value],
        };

        return acc;
      }
    }

    acc[currentRuleKey] = rule;

    return acc;
  }, {});

  return Object.values(groupedRulesMap);
}
//#endregion Private Members
