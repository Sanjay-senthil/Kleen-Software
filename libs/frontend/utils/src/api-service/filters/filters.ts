import {
  FilterCombinator,
  FilterOperation,
  FilterOperators,
  FilterQuery,
  FilterRule,
  Filters,
  Maybe,
  ParamsBasedOnRoute,
  PrimitiveType,
} from '@kleeen/types';
import { isNilOrEmpty, isPrimitive } from '@kleeen/common/utils';

import { INVESTIGATION_URL_PARAM } from '@kleeen/investigations';
import camelCase from 'lodash.camelcase';

const operationMap: { [key in FilterOperators]: FilterOperation } = {
  [FilterOperators.from]: FilterOperation.GreaterOrEqualThan,
  [FilterOperators.in]: FilterOperation.Is,
  [FilterOperators.max]: FilterOperation.LessOrEqualThan,
  [FilterOperators.min]: FilterOperation.GreaterOrEqualThan,
  [FilterOperators.notIn]: FilterOperation.IsNot,
  [FilterOperators.relativeDate]: FilterOperation.RelativeDate,
  [FilterOperators.to]: FilterOperation.LessOrEqualThan,
};

export function getFiltersInput(filters?: Filters): FilterQuery | Record<string, never> {
  if (isNilOrEmpty(filters)) return {};

  const rules = Object.entries(filters).reduce((acc: FilterRule[], [thingName, operation]) => {
    if (thingName === INVESTIGATION_URL_PARAM) {
      // Ignore the special Investigation param
      return acc;
    }

    const field = getFilterFieldName(thingName);

    if (isPrimitive(operation)) {
      const ruleFromPrimitive = {
        field,
        operation: FilterOperation.Is,
        value: { id: operation },
      };

      acc.push(ruleFromPrimitive);

      return acc;
    }

    const filterRulesFromObject = getFilterOperationsAndValues({
      field,
      operation,
    });

    if (isNilOrEmpty(filterRulesFromObject)) return acc;

    acc.push(...filterRulesFromObject);

    return acc;
  }, []);
  const filtersInput = {
    combinator: FilterCombinator.AND,
    rules,
  };

  return filtersInput;
}

//#region Private members
interface GetFilterOperationsAndValues {
  field: string;
  operation: ParamsBasedOnRoute;
}

function getFilterOperationsAndValues({
  field,
  operation,
}: GetFilterOperationsAndValues): Maybe<FilterRule[]> {
  if (isNilOrEmpty(operation)) return;

  const filterOperationsAndValues = Object.entries(operation).reduce(
    (acc: FilterRule[], [oldOperator, value]) => {
      if (isNilOrEmpty(value)) return acc;

      const baseFilterOperationAndValue = {
        field,
        operation: operationMap[oldOperator as FilterOperators],
      };

      if (Array.isArray(value)) {
        const displayValues = value as PrimitiveType[];
        const rules = displayValues.map((displayValue): FilterRule => {
          return {
            ...baseFilterOperationAndValue,
            value: {
              displayValue,
            },
          };
        });

        acc.push(...rules);

        return acc;
      }

      const displayValue = value as PrimitiveType;
      const filterOperationAndValue = {
        ...baseFilterOperationAndValue,
        value: { displayValue },
      };

      acc.push(filterOperationAndValue);

      return acc;
    },
    [],
  );

  return filterOperationsAndValues;
}

function getFilterFieldName(originalName: string): string {
  return camelCase(originalName);
}
//#endregion
