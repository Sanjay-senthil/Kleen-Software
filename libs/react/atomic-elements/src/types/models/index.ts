import { CrossLinking, FormatProps, ResultsProps } from '@kleeen/types';

export interface ContextProps {
  data: DataProps;
  error: string;
  isLoading: boolean;
}

export interface DataProps extends ResultsProps {
  crossLinking: CrossLinking[] | CrossLinking;
  format: FormatProps;
}

export interface XAxisOptions extends Highcharts.XAxisOptions {
  categories?: [string];
  key?: string;
  valueLabels?: { label: string; value: number }[];
}

export interface YAxisOptions extends Highcharts.YAxisOptions {
  categories?: [string];
  key?: string;
  valueLabels?: { label: string; value: number }[];
}
