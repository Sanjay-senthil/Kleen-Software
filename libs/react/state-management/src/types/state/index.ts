export * from './endUser';
export * from './globalAppErrors';
export * from './investigation';
