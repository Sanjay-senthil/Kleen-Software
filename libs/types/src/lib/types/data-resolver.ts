import {
  Bucket,
  CrossLinking,
  CrossLinkingMatrix,
  DisplayValue,
  Transformation,
  UpdateDataStrategy,
} from '@kleeen/types';

/**
 * Enumerations
 */

export enum MapFunctionType {
  ToDisplayValue = 'toDisplayValue',
  ToChipsValue = 'toChipsValue',
  ToIdValue = 'toIdValue',
}

/**
 * Interfaces
 */

export interface Axis {
  categories?: string[] | ListItemBase[];
  key?: string;
  type?: string;
}

export interface ColumnMap {
  handler?: MapFunctionType;
  idPath?: string;
  key: string;
  path: string;
  showDisplayMedia?: boolean;
}

export interface Entity {
  name: string;
  properties: { [property: string]: { statisticalType: string } };
}

export interface EntityList {
  [key: string]: ListingItem | ListingItem[];
}

export interface DataAggregationArgsDataPoint {
  bucket?: Bucket;
  name: string;
  transformation?: Transformation;
  transformations?: Transformation[];
}

export interface GetListingDataResults {
  data: EntityList[];
  format: { [key: string]: ListingFormatItem };
  latestRequestTimestamp?: number;
  pagination?: { totalCount: number };
  strategy?: UpdateDataStrategy;
}

export interface GetWidgetDataResult {
  crossLinking?: CrossLinkingMatrix;
  format: {
    xAxis?: Axis;
    yAxis?: Axis;
  };
  results: number[] | number[][];
}

export interface ListingFormatItem {
  aggregations?: null;
  examples?: null;
  max?: null;
  min?: null;
  prefix?: null;
  severityBad?: null;
  severityGood?: null;
  severityLevels?: null;
  suffix?: null;
}

export interface ListingItem {
  displayValue: DisplayValue;
  id?: string;
}

export interface MultiTransFormationResults {
  crossLinking?: CrossLinking[];
  format: {
    xAxis?: Axis;
    yAxis?: Axis;
  };
  results: number | number[] | number[][];
  transformation: Transformation;
}

export interface TransformationArgs extends Pick<ColumnMap, 'idPath' | 'key' | 'path' | 'showDisplayMedia'> {
  row: any;
}

/**
 * Types
 */

export type ListItemBase = number | string | boolean | undefined;
