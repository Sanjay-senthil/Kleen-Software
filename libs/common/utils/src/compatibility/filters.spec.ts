/* eslint-disable max-lines */
import { FilterCombinator, FilterOperation, FilterOperators, FilterQuery, IntervalDate } from '@kleeen/types';

import { transformFiltersToDeprecatedVersion } from './filters';

describe('compatibility', () => {
  describe('filters', () => {
    it('should return undefined if "filters" is nil', () => {
      expect(
        transformFiltersToDeprecatedVersion({
          filters: undefined,
        } as { filters: FilterQuery }),
      ).toBeUndefined();
      expect(
        transformFiltersToDeprecatedVersion({
          filters: null,
        } as { filters: FilterQuery }),
      ).toBeUndefined();
      expect(
        transformFiltersToDeprecatedVersion({
          filters: {},
        } as { filters: FilterQuery }),
      ).toBeUndefined();
    });

    it('should return undefined if "filters" has no rules', () => {
      const filters = {
        combinator: FilterCombinator.AND,
      };

      expect(
        transformFiltersToDeprecatedVersion({
          filters,
        } as { filters: FilterQuery }),
      ).toBeUndefined();
    });

    it('should return an empty filters object if "filters" has empty rules', () => {
      const filters = {
        combinator: FilterCombinator.AND,
        rules: [],
      };
      const filtersDeprecatedVersion = {};

      expect(
        transformFiltersToDeprecatedVersion({
          filters,
        }),
      ).toEqual(filtersDeprecatedVersion);
    });

    it(`should return a filters object with three "plain" keys if "filters" has three "${FilterOperation.Is}" rules`, () => {
      const filters = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'computerChild',
            operation: FilterOperation.Is,
            value: { id: '39a91f5e-da34-4e64-bd0f-730a7c9d6a99' },
          },
          {
            field: 'computerParent',
            operation: FilterOperation.Is,
            value: { id: 12345 },
          },
          {
            field: 'isParent',
            operation: FilterOperation.Is,
            value: { id: false },
          },
        ],
      };
      const filtersDeprecatedVersion = {
        computerChild: '39a91f5e-da34-4e64-bd0f-730a7c9d6a99',
        computerParent: 12345,
        isParent: false,
      };

      expect(
        transformFiltersToDeprecatedVersion({
          filters,
        }),
      ).toEqual(filtersDeprecatedVersion);
    });

    it(`should return a filters object with three "${FilterOperators.in}" keys if "filters" has four "${FilterOperation.Is}" rules and two of them belong to the same field`, () => {
      const filters = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'computer',
            operation: FilterOperation.Is,
            value: { displayValue: 12345 },
          },
          {
            field: 'computerChild',
            operation: FilterOperation.Is,
            value: { displayValue: 'quia aut dolorum at porro deserunt accusamus' },
          },
          {
            field: 'computerParent',
            operation: FilterOperation.Is,
            value: { displayValue: 'optio nobis maiores' },
          },
          {
            field: 'computerParent',
            operation: FilterOperation.Is,
            value: { displayValue: 'quisquam voluptate molestias eius' },
          },
        ],
      };
      const filtersDeprecatedVersion = {
        Computer: {
          [FilterOperators.in]: [12345],
        },
        ComputerChild: {
          [FilterOperators.in]: ['quia aut dolorum at porro deserunt accusamus'],
        },
        ComputerParent: {
          [FilterOperators.in]: ['optio nobis maiores', 'quisquam voluptate molestias eius'],
        },
      };

      expect(
        transformFiltersToDeprecatedVersion({
          filters,
        }),
      ).toEqual(filtersDeprecatedVersion);
    });

    it(`should return a filters object with three "${FilterOperators.notIn}" keys if "filters" has four "${FilterOperation.IsNot}" rules and two of them belong to the same field`, () => {
      const filters = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'computer',
            operation: FilterOperation.IsNot,
            value: { displayValue: 12345 },
          },
          {
            field: 'computerChild',
            operation: FilterOperation.IsNot,
            value: { displayValue: 'quia aut dolorum at porro deserunt accusamus' },
          },
          {
            field: 'computerParent',
            operation: FilterOperation.IsNot,
            value: { displayValue: 'optio nobis maiores' },
          },
          {
            field: 'computerParent',
            operation: FilterOperation.IsNot,
            value: { displayValue: 'quisquam voluptate molestias eius' },
          },
        ],
      };
      const filtersDeprecatedVersion = {
        Computer: {
          [FilterOperators.notIn]: [12345],
        },
        ComputerChild: {
          [FilterOperators.notIn]: ['quia aut dolorum at porro deserunt accusamus'],
        },
        ComputerParent: {
          [FilterOperators.notIn]: ['optio nobis maiores', 'quisquam voluptate molestias eius'],
        },
      };

      expect(
        transformFiltersToDeprecatedVersion({
          filters,
        }),
      ).toEqual(filtersDeprecatedVersion);
    });

    it(`should return a filters object with one "${FilterOperators.max}" key if "filters" has one "${FilterOperation.LessOrEqualThan}" rule`, () => {
      const filters = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'salary',
            operation: FilterOperation.LessOrEqualThan,
            value: { displayValue: 80 },
          },
        ],
      };
      const filtersDeprecatedVersion = {
        Salary: {
          [FilterOperators.max]: 80,
        },
      };

      expect(
        transformFiltersToDeprecatedVersion({
          filters,
        }),
      ).toEqual(filtersDeprecatedVersion);
    });

    it(`should return a filters object with one "${FilterOperators.min}" key if "filters" has one "${FilterOperation.GreaterOrEqualThan}" rule`, () => {
      const filters = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'salary',
            operation: FilterOperation.GreaterOrEqualThan,
            value: { displayValue: -45 },
          },
        ],
      };
      const filtersDeprecatedVersion = {
        Salary: {
          [FilterOperators.min]: -45,
        },
      };

      expect(
        transformFiltersToDeprecatedVersion({
          filters,
        }),
      ).toEqual(filtersDeprecatedVersion);
    });

    it(`should return a filters object with one "${FilterOperators.min}" and one "${FilterOperators.max}" keys if "filters" has one "${FilterOperation.GreaterOrEqualThan}" and one "${FilterOperation.LessOrEqualThan}" rules`, () => {
      const filters = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'salary',
            operation: FilterOperation.LessOrEqualThan,
            value: { displayValue: 80 },
          },
          {
            field: 'salary',
            operation: FilterOperation.GreaterOrEqualThan,
            value: { displayValue: -45 },
          },
        ],
      };
      const filtersDeprecatedVersion = {
        Salary: {
          [FilterOperators.max]: 80,
          [FilterOperators.min]: -45,
        },
      };

      expect(
        transformFiltersToDeprecatedVersion({
          filters,
        }),
      ).toEqual(filtersDeprecatedVersion);
    });

    it(`should return a filters object with one "${FilterOperators.from}" and one "${FilterOperators.to}" keys if "filters" has one "${FilterOperation.GreaterOrEqualThan}" and one "${FilterOperation.LessOrEqualThan}" timestamp rules`, () => {
      const filters = {
        combinator: FilterCombinator.AND,
        rules: [
          {
            field: 'timestamp',
            operation: FilterOperation.GreaterOrEqualThan,
            value: { displayValue: 1630497600000 },
          },
          {
            field: 'timestamp',
            operation: FilterOperation.LessOrEqualThan,
            value: { displayValue: 1632794400000 },
          },
        ],
      };
      const filtersDeprecatedVersion = {
        Timestamp: {
          [FilterOperators.from]: 1630497600000,
          [FilterOperators.to]: 1632794400000,
        },
      };

      expect(
        transformFiltersToDeprecatedVersion({
          filters,
        }),
      ).toEqual(filtersDeprecatedVersion);
    });

    it.each([
      [IntervalDate.allTime, IntervalDate.allTime],
      [IntervalDate.minute, IntervalDate.minute],
      [IntervalDate.oneHours, IntervalDate.oneHours],
      [IntervalDate.oneMonth, IntervalDate.oneMonth],
      [IntervalDate.oneWeek, IntervalDate.oneWeek],
      [IntervalDate.sixHours, IntervalDate.sixHours],
      [IntervalDate.threeMonth, IntervalDate.threeMonth],
      [IntervalDate.twentyFourHours, IntervalDate.twentyFourHours],
    ])(
      `should return a filters object with one %p "${FilterOperators.relativeDate}" key if "filters" has one %p "${FilterOperation.RelativeDate}" rule`,
      (interval, expectedInterval) => {
        const filters = {
          combinator: FilterCombinator.AND,
          rules: [
            {
              field: 'timestamp',
              operation: FilterOperation.RelativeDate,
              value: { displayValue: interval },
            },
          ],
        };
        const filtersDeprecatedVersion = {
          Timestamp: {
            [FilterOperators.relativeDate]: expectedInterval,
          },
        };

        expect(
          transformFiltersToDeprecatedVersion({
            filters,
          }),
        ).toEqual(filtersDeprecatedVersion);
      },
    );

    it(`should return a filters objects with all keys if "filters" has all operators`, () => {
      const filters = {
        combinator: FilterCombinator.AND,
        rules: [
          { field: 'coffeeShop', operation: FilterOperation.Is, value: { displayValue: 'Chocolatito' } },
          { field: 'salary', operation: FilterOperation.Is, value: { displayValue: 20 } },
          { field: 'salary', operation: FilterOperation.GreaterOrEqualThan, value: { displayValue: 14 } },
          { field: 'salary', operation: FilterOperation.LessOrEqualThan, value: { displayValue: 23 } },
          {
            field: 'timestamp',
            operation: FilterOperation.GreaterOrEqualThan,
            value: { displayValue: 1633101300000 },
          },
          {
            field: 'timestamp',
            operation: FilterOperation.LessOrEqualThan,
            value: { displayValue: 1633274100000 },
          },
        ],
      };

      expect(
        transformFiltersToDeprecatedVersion({
          filters,
        }),
      ).toMatchSnapshot();
    });
  });
});
