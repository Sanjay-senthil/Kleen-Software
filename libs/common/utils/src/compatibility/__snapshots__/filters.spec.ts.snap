// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compatibility filters should return a filters objects with all keys if "filters" has all operators 1`] = `
Object {
  "CoffeeShop": Object {
    "_in": Array [
      "Chocolatito",
    ],
  },
  "Salary": Object {
    "_in": Array [
      20,
    ],
    "max": 23,
    "min": 14,
  },
  "Timestamp": Object {
    "from": 1633101300000,
    "to": 1633274100000,
  },
}
`;
